addons/stock_account/wizard/stock_valuation_layer_revaluation.py

```diff
@@ -0,0 +1,135 @@
+    product_uom_name = fields.Char('Unit of Measure', related='product_id.uom_id.name')
+    current_value_svl = fields.Float("Current Value", related="product_id.value_svl")
+    current_quantity_svl = fields.Float("Current Quantity", related="product_id.quantity_svl")
+
+    added_value = fields.Monetary("Added value", required=True)
+    new_value = fields.Monetary("New value", compute='_compute_new_value')
+    new_value_by_qty = fields.Monetary("New value by quantity", compute='_compute_new_value')
+    reason = fields.Char("Reason", help="Reason of the revaluation")
+
+    account_journal_id = fields.Many2one('account.journal', "Journal", required=True, check_company=True)
+    account_id = fields.Many2one('account.account', "Account", domain=[('deprecated', '=', False)], required=True, check_company=True)
+    date = fields.Date("Account Move Date")
+
+    @api.depends('current_value_svl', 'current_quantity_svl', 'added_value')
+    def _compute_new_value(self):
+        for revalue in self:
+            revalue.new_value = revalue.current_value_svl + revalue.added_value
+            if not self.company_id.currency_id.is_zero(revalue.current_quantity_svl):
+                revalue.new_value_by_qty = revalue.new_value / revalue.current_quantity_svl
+            else:
+                revalue.new_value_by_qty = 0.0
+
+    def action_validate_revaluation(self):
+        """ Revaluate the stock for `self.product_id` in `self.company_id`.
+
+        - Change the stardard price with the new valuation by product unit.
+        - Create a manual stock valuation layer with the `added_value` of `self`.
+        - If the Inventory Valuation of the product category is automated, create
+        related account move.
+        """
+        self.ensure_one()
+        if self.company_id.currency_id.is_zero(self.added_value):
+            raise UserError(_("The added value doesn't have any impact after rounding"))
+
+        slvs_to_update = self.env['stock.valuation.layer'].with_company(self.company_id).search([
+            ('product_id', '=', self.product_id.id),
+            ('remaining_qty', '>', 0),
+            ('company_id', '=', self.company_id.id),
+        ])
+
+        # Create a manual stock valuation layer
+        svl_vals = {
+            'company_id': self.company_id.id,
+            'product_id': self.product_id.id,
+            'description': _("Manual Stock Valuation: %s.%s") % (
+                self.reason or _("No Reason Given"),
+                self.product_id.categ_id.property_cost_method == 'average' and (_(" Product cost move from %s to %s") % (self.product_id.standard_price, self.new_value_by_qty)) or ''),
+            'value': self.added_value,
+            'quantity': 0,
+        }
+
+        ctn_remaining_quantity = sum(slv.remaining_qty for slv in slvs_to_update)
+        for slv in slvs_to_update:
+            slv.remaining_value += self.added_value * slv.remaining_qty / ctn_remaining_quantity
```

Comment:
if there's 3 remaining layer and you want to add $1 of value, you'll create a layer of value $1 and add on the remaining_layers $0.33 $0.33 $0.33